### ISSUE TITLE
"-d" fails with "basic_filebuf::underflow error reading the file"

### ISSUE THREAD

**Issue by solardiz** [kimci86/bkcrack#4]
Using the provided example, I get this error:

```
[solar@super build]$ src/bkcrack -C ../example/secrets.zip -c spiral.svg -k c4038591 d5ff449d d3b0c696 -d spiral_deciphered.svg
[20:34:01] Keys
c4038591 d5ff449d d3b0c696
terminate called after throwing an instance of 'std::ios_base::failure'
  what():  basic_filebuf::underflow error reading the file
Aborted
```

strace shows:

```
open("../example/secrets.zip", O_RDONLY) = 3
lseek(3, -22, SEEK_END)                 = 56241
read(3, "PK\5\6\0\0\0\0\2\0\2\0\270\0\0\0\371\332\0\0\0\0", 8191) = 22
lseek(3, -22, SEEK_CUR)                 = 56241
lseek(3, 0, SEEK_CUR)                   = 56241
lseek(3, 16, SEEK_CUR)                  = 56257
read(3, "\371\332\0\0\0\0", 8191)       = 6
lseek(3, 56057, SEEK_SET)               = 56057
lseek(3, 0, SEEK_CUR)                   = 56057
lseek(3, 20, SEEK_CUR)                  = 56077
read(3, "\254\325\0\0\17\326\0\0\n\0$\0\0\0\0\0\0\0 \200\264\201\0\0\0\0advice"..., 8191) = 186
lseek(3, -178, SEEK_CUR)                = 56085
read(3, "\n\0$\0\0\0\0\0\0\0 \200\264\201\0\0\0\0advice.jpg\n\0 \0"..., 8191) = 178
lseek(3, -164, SEEK_CUR)                = 56099
read(3, "\0\0\0\0advice.jpg\n\0 \0\0\0\0\0\1\0\30\0\0\224\231\330#z"..., 8191) = 164
lseek(3, -114, SEEK_CUR)                = 56149
lseek(3, 0, SEEK_CUR)                   = 56149
lseek(3, 20, SEEK_CUR)                  = 56169
read(3, "\375\4\0\0\361\4\0\0\n\0$\0\0\0\0\0\0\0 \200\264\201\324\325\0\0spiral"..., 8191) = 94
lseek(3, -86, SEEK_CUR)                 = 56177
read(3, "\n\0$\0\0\0\0\0\0\0 \200\264\201\324\325\0\0spiral.svg\n\0 \0"..., 8191) = 86
lseek(3, -72, SEEK_CUR)                 = 56191
read(3, "\324\325\0\0spiral.svg\n\0 \0\0\0\0\0\1\0\30\0\0\213wB`\230"..., 8191) = 72
lseek(3, -22, SEEK_CUR)                 = 56241
lseek(3, 54768, SEEK_SET)               = 54768
read(3, "\0\0spiral.svg\325\32\320\311~P\353\370\336(-$Nd\377\20+\213\242\224"..., 8191) = 1495
lseek(3, -1483, SEEK_CUR)               = 54780
open("spiral_deciphered.svg", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 4
read(3, NULL, 8191)                     = -1 EFAULT (Bad address)
futex(0x7f4c63611b10, FUTEX_WAKE_PRIVATE, 2147483647) = 0
write(2, "terminate called after throwing "..., 48terminate called after throwing an instance of ') = 48
```

So it's trying to read(2) into a NULL pointer, and this happens right after it's opened the output file. This corresponds to somewhere in:

```c
        // discard the encryption header
        std::istreambuf_iterator<char> cipher(cipherstream);
        std::size_t i;
        for(i = 0; i < Data::headerSize && cipher != std::istreambuf_iterator<char>(); i++, ++cipher)
            keys.update(*cipher ^ KeystreamTab::getByte(keys.getZ()));

        for(std::ostreambuf_iterator<char> plain(decipheredstream); i < ciphersize && cipher != std::istreambuf_iterator<char>(); i++, ++cipher, ++plain)
        {
            byte p = *cipher ^ KeystreamTab::getByte(keys.getZ());
            keys.update(p);
            *plain = p;
        }
```

The system is Scientific Linux 6.10 with cmake3 and several devtoolset's installed to provide newer development tools. I've tried rebuilding with devtoolsets 2 through 6, which provide gcc versions 4.8.2 to 6.3.1 (I originally started with the latest), which didn't make a difference.

If you need, I can provide remote access to this system, see https://openwall.info/wiki/HPC/Village

**Comment by kimci86** [comment]
It might be related to [a bug in GCC](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53984).
Can you use a newer version of GCC in your environment?
Not sure how to workaround this problem.

**Comment by solardiz** [comment]
I've just installed devtoolset-7, which provides "gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)", but got the exact same error after rebuilding with that version. The GCC bug you reference mentions backports of the fixes "for 5.5, 6.5 and 7.3" in 2017, so it would seem a 7.3.x release from 2018 would already have the fix, but maybe not.

I could build a newer gcc from source and test with that, but anyhow with the issue present across such a wide range of gcc versions still in use by distros a workaround in bkcrack is needed or a lot of people would not be able to easily use bkcrack for a long time.

**Comment by kimci86** [comment]
I hunted the bug on your environment, I can confirm the bug in GCC referenced earlier is still present even with devtoolset-7. (I tested it with a small example given in GCC's bugzilla ticket.)
Anyway it seems to me the issue happens in bkcrack because an ifstream is move-assigned. Apparently the implementation on this environment does not like that. I will write a fix to work around this.
Thanks for reporting the issue!
